// @ngstack/code-editor/esm5/lib/services/javascript-defaults.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { CodeEditorService, TypingsInfo } from './code-editor.service';
import * as i0 from "@angular/core";
import * as i1 from "./code-editor.service";
var JavascriptDefaultsService = /** @class */ (function () {
    function JavascriptDefaultsService(codeEditorService) {
        var _this = this;
        codeEditorService.loaded.subscribe(function (event) {
            _this.setup(event.monaco);
        });
        codeEditorService.typingsLoaded.subscribe(function (typings) {
            _this.updateTypings(typings);
        });
    }
    JavascriptDefaultsService.prototype.setup = function (monaco) {
        if (!monaco) {
            return;
        }
        this.monaco = monaco;
        var defaults = monaco.languages.typescript.javascriptDefaults;
        defaults.setCompilerOptions({
            target: monaco.languages.typescript.ScriptTarget.ES6,
            module: 'commonjs',
            allowNonTsExtensions: true,
            baseUrl: '.',
            paths: {}
        });
        defaults.setMaximumWorkerIdleTime(-1);
        defaults.setEagerModelSync(true);
        /*
        defaults.setDiagnosticsOptions({
          noSemanticValidation: false,
          noSyntaxValidation: false
        });
        */
    };
    JavascriptDefaultsService.prototype.updateTypings = function (typings) {
        if (typings) {
            this.addExtraLibs(typings.files);
            this.addLibraryPaths(typings.entryPoints);
        }
    };
    JavascriptDefaultsService.prototype.addExtraLibs = function (libs) {
        if (libs === void 0) { libs = []; }
        if (!this.monaco || !libs || libs.length === 0) {
            return;
        }
        var defaults = this.monaco.languages.typescript.javascriptDefaults;
        // undocumented API
        var registeredLibs = defaults.getExtraLibs();
        libs.forEach(function (lib) {
            if (!registeredLibs[lib.path]) {
                // needs performance improvements, recreates its worker each time
                // defaults.addExtraLib(lib.content, lib.path);
                // undocumented API
                defaults._extraLibs[lib.path] = lib.content;
            }
        });
        // undocumented API
        defaults._onDidChange.fire(defaults);
    };
    JavascriptDefaultsService.prototype.addLibraryPaths = function (paths) {
        if (paths === void 0) { paths = {}; }
        if (!this.monaco) {
            return;
        }
        var defaults = this.monaco.languages.typescript.javascriptDefaults;
        var compilerOptions = defaults.getCompilerOptions();
        compilerOptions.paths = compilerOptions.paths || {};
        Object.keys(paths).forEach(function (key) {
            compilerOptions.paths[key] = [paths[key]];
        });
    };
    JavascriptDefaultsService.ctorParameters = function () { return [
        { type: CodeEditorService }
    ]; };
    JavascriptDefaultsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function JavascriptDefaultsService_Factory() { return new JavascriptDefaultsService(i0.ɵɵinject(i1.CodeEditorService)); }, token: JavascriptDefaultsService, providedIn: "root" });
    JavascriptDefaultsService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [CodeEditorService])
    ], JavascriptDefaultsService);
    return JavascriptDefaultsService;
}());
export { JavascriptDefaultsService };
