// @ngstack/code-editor/esm5/lib/services/json-defaults.service.js
import { __decorate, __metadata, __read, __spread, __values } from "tslib";
import { Injectable } from '@angular/core';
import { CodeEditorService } from './code-editor.service';
import * as i0 from "@angular/core";
import * as i1 from "./code-editor.service";
var JsonDefaultsService = /** @class */ (function () {
    function JsonDefaultsService(codeEditorService) {
        var _this = this;
        codeEditorService.loaded.subscribe(function (event) {
            _this.setup(event.monaco);
        });
    }
    JsonDefaultsService.prototype.setup = function (monaco) {
        if (!monaco) {
            return;
        }
        this.monaco = monaco;
        var defaults = monaco.languages.json.jsonDefaults;
        defaults.setDiagnosticsOptions({
            validate: true,
            allowComments: true,
            schemas: __spread(defaults._diagnosticsOptions.schemas, [
                {
                    uri: 'http://myserver/foo-schema.json',
                    // fileMatch: [id],
                    // fileMatch: ['*.json'],
                    schema: {
                        type: 'object',
                        properties: {
                            p1: {
                                enum: ['v1', 'v2']
                            },
                            p2: {
                                $ref: 'http://myserver/bar-schema.json'
                            }
                        }
                    }
                },
                {
                    uri: 'http://myserver/bar-schema.json',
                    // fileMatch: [id],
                    // fileMatch: ['*.json'],
                    schema: {
                        type: 'object',
                        properties: {
                            q1: {
                                enum: ['x1', 'x2']
                            }
                        }
                    }
                }
            ])
        });
    };
    JsonDefaultsService.prototype.addSchemas = function (id, definitions) {
        var e_1, _a;
        if (definitions === void 0) { definitions = []; }
        var defaults = this.monaco.languages.json.jsonDefaults;
        var options = defaults.diagnosticsOptions;
        var schemas = {};
        if (options && options.schemas && options.schemas.length > 0) {
            options.schemas.forEach(function (schema) {
                schemas[schema.uri] = schema;
            });
        }
        try {
            for (var definitions_1 = __values(definitions), definitions_1_1 = definitions_1.next(); !definitions_1_1.done; definitions_1_1 = definitions_1.next()) {
                var _b = definitions_1_1.value, uri = _b.uri, schema = _b.schema;
                schemas[uri] = {
                    uri: uri,
                    schema: schema,
                    fileMatch: [id || '*.json']
                };
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (definitions_1_1 && !definitions_1_1.done && (_a = definitions_1.return)) _a.call(definitions_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        // console.log(schemas);
        // console.log(Object.values(schemas));
        options.schemas = Object.values(schemas);
        defaults.setDiagnosticsOptions(options);
    };
    JsonDefaultsService.ctorParameters = function () { return [
        { type: CodeEditorService }
    ]; };
    JsonDefaultsService.ɵprov = i0.ɵɵdefineInjectable({ factory: function JsonDefaultsService_Factory() { return new JsonDefaultsService(i0.ɵɵinject(i1.CodeEditorService)); }, token: JsonDefaultsService, providedIn: "root" });
    JsonDefaultsService = __decorate([
        Injectable({
            providedIn: 'root'
        }),
        __metadata("design:paramtypes", [CodeEditorService])
    ], JsonDefaultsService);
    return JsonDefaultsService;
}());
export { JsonDefaultsService };
