// @ngstack/code-editor/esm2015/lib/code-editor.module.js
var CodeEditorModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CodeEditorComponent } from './code-editor/code-editor.component';
import { CodeEditorService, EDITOR_SETTINGS } from './services/code-editor.service';
import { TypescriptDefaultsService } from './services/typescript-defaults.service';
import { JavascriptDefaultsService } from './services/javascript-defaults.service';
import { JsonDefaultsService } from './services/json-defaults.service';
let CodeEditorModule = CodeEditorModule_1 = class CodeEditorModule {
    static forRoot(settings) {
        return {
            ngModule: CodeEditorModule_1,
            providers: [
                { provide: EDITOR_SETTINGS, useValue: settings },
                CodeEditorService,
                TypescriptDefaultsService,
                JavascriptDefaultsService,
                JsonDefaultsService
            ]
        };
    }
    static forChild() {
        return {
            ngModule: CodeEditorModule_1
        };
    }
};
CodeEditorModule = CodeEditorModule_1 = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [CodeEditorComponent],
        exports: [CodeEditorComponent]
    })
], CodeEditorModule);
export { CodeEditorModule };
