// ngx-monaco-editor/esm5/lib/editor.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NGX_MONACO_EDITOR_CONFIG } from './config';
import { DiffEditorComponent } from './diff-editor.component';
import { EditorComponent } from './editor.component';
var MonacoEditorModule = /** @class */ (function () {
    function MonacoEditorModule() {
    }
    MonacoEditorModule_1 = MonacoEditorModule;
    MonacoEditorModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: MonacoEditorModule_1,
            providers: [
                { provide: NGX_MONACO_EDITOR_CONFIG, useValue: config }
            ]
        };
    };
    var MonacoEditorModule_1;
    MonacoEditorModule = MonacoEditorModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule
            ],
            declarations: [
                EditorComponent,
                DiffEditorComponent
            ],
            exports: [
                EditorComponent,
                DiffEditorComponent
            ]
        })
    ], MonacoEditorModule);
    return MonacoEditorModule;
}());
export { MonacoEditorModule };
