// @ngstack/code-editor/esm5/lib/code-editor.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CodeEditorComponent } from './code-editor/code-editor.component';
import { CodeEditorService, EDITOR_SETTINGS } from './services/code-editor.service';
import { TypescriptDefaultsService } from './services/typescript-defaults.service';
import { JavascriptDefaultsService } from './services/javascript-defaults.service';
import { JsonDefaultsService } from './services/json-defaults.service';
var CodeEditorModule = /** @class */ (function () {
    function CodeEditorModule() {
    }
    CodeEditorModule_1 = CodeEditorModule;
    CodeEditorModule.forRoot = function (settings) {
        return {
            ngModule: CodeEditorModule_1,
            providers: [
                { provide: EDITOR_SETTINGS, useValue: settings },
                CodeEditorService,
                TypescriptDefaultsService,
                JavascriptDefaultsService,
                JsonDefaultsService
            ]
        };
    };
    CodeEditorModule.forChild = function () {
        return {
            ngModule: CodeEditorModule_1
        };
    };
    var CodeEditorModule_1;
    CodeEditorModule = CodeEditorModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [CodeEditorComponent],
            exports: [CodeEditorComponent]
        })
    ], CodeEditorModule);
    return CodeEditorModule;
}());
export { CodeEditorModule };
